services/src/main/java/org/keycloak/vault/PlainTextVaultProvider.java

```diff
@@ -0,0 +1,84 @@
+package org.keycloak.vault;
+
+import org.jboss.logging.Logger;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.MappedByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.EnumSet;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+/**
+ * A text-based vault provider, which stores each secret in a separate file. The file name needs to match a
+ * vault secret id (or a key for short). A typical vault directory layout looks like this:
+ * <pre>
+ *     ${VAULT}/realma__key1 (contains secret for key 1)
+ *     ${VAULT}/realma__key2 (contains secret for key 2)
+ *     etc...
+ * </pre>
+ * Note, that each key needs is prefixed by realm name. This kind of layout is used by Kubernetes by default
+ * (when mounting a volume into the pod).
+ *
+ * See https://kubernetes.io/docs/concepts/configuration/secret/
+ * See https://github.com/keycloak/keycloak-community/blob/master/design/secure-credentials-store.md#plain-text-file-per-secret-kubernetes--openshift
+ *
+ * @author Sebastian Łaskawiec
+ */
+public class PlainTextVaultProvider implements VaultProvider {
+
+    private static final Logger logger = Logger.getLogger(MethodHandles.lookup().lookupClass());
+
+    private final Path vaultPath;
+    private final BiFunction<Path, String, Path> secretMappingFunction;
+
+    /**
+     * Creates a new {@link PlainTextVaultProvider}.
+     *
+     * @param path A path to a vault. Can not be null.
+     * @param realmName A realm name. Can not be null.
+     */
+    public PlainTextVaultProvider(@Nonnull Path path, @Nonnull String realmName) {
+        vaultPath = path;
+        secretMappingFunction = (vaultDirectory, secretId) -> vaultDirectory.resolve(realmName + "__" + secretId);
```

Comment:
a) Is the anonymous method really necessary here for the initial implementation? Why not use simple `private final String prefix = realmName.replaceAll("_", "__") + "_"` field instead?
b) The function to obtain path from realm name to secret needs to be in line with the design. Currently it is other way round: it does not escape underscores, and it adds two additional ones. It would not work for realm with double underscore in it.

> "Every occurence of an underscore in both realm name and secret key is replaced with two underscores, so e.g. for realm realm_c and Keycloak key smtp_password, the secret key (and thus file name) would be realm__c_smtp__password."

Reply:
> a) Is the anonymous method really necessary here for the initial implementation? Why not use simple private final String prefix = realmName.replaceAll("_", "__") + "_" field instead?

@hmlnarik This is very flexible approach I think, where the caller can define whatever mapping he wants. It's also very useful for tests, where I'm using a temporary files generated by the JVM+OS and I have very little control on the name of this file. See https://github.com/keycloak/keycloak/pull/6223/files#diff-3b6b8f2966e0b136c01bbbe19263670cR97. I would prefer to leave it as is - I think it's useful.

> b) The function to obtain path from realm name to secret needs to be in line with the design. Currently it is other way round: it does not escape underscores, and it adds two additional ones. It would not work for realm with double underscore in it.

@hmlnarik Ohhh yes, now I see the design. Sorry for that, for some reason I must have read it in a wrong way.

Reply:
@hmlnarik Should be fine now. I also added a test. If you're happy, please resolve this conversation.